base/REPL.jl

```diff
@@ -70,28 +82,70 @@ function eval_user_input(ast::ANY, backend::REPLBackend)
                 println("SYSTEM ERROR: Failed to report error to REPL frontend")
                 println(err)
             end
-            iserr, lasterr = true, err
-            bt = catch_backtrace()
+            iserr, lasterr = true, (err, catch_backtrace())
         end
     end
 end
 
+"""    set_unhandled_exception_handler(f)
+Register the function `f` as the current unhandled exception handler.
+If `f` is not a function, the handler is uninstalled.
+The return value is the current exception handler, or `nothing` if none is installed.
+"""
+function set_unhandled_exception_handler(f)
+    ccall(:jl_set_unhandled_exception_handler, Any, (Any,), f)
+end
+
 function start_repl_backend(repl_channel::Channel, response_channel::Channel)
-    backend = REPLBackend(repl_channel, response_channel, false, nothing)
-    backend.backend_task = @schedule begin
+    local backend # keep track of the current backend
+                  # each backend can look at this to see if it has been replaced
+                  # by the unhandled_exception_handler
+    new_backend_task = function()
         # include looks at this to determine the relative include path
         # nothing means cwd
-        while true
+        while current_task() == backend.task
             tls = task_local_storage()
             tls[:SOURCE_PATH] = nothing
-            ast, show_value = take!(backend.repl_channel)
+            ast, show_value = take!(repl_channel)
             if show_value == -1
-                # exit flag
+                # graceful exit flag
+                set_unhandled_exception_handler(last)
                 break
             end
+            backend.responded = false
             eval_user_input(ast, backend)
         end
+        nothing
     end
+    backend = REPLBackend(repl_channel, response_channel, Task(new_backend_task), true)
+    schedule(backend.task)
+
+    last = set_unhandled_exception_handler() do excpt
+        if istaskdone(backend.task)
+            # our backend is gone, so reset to the previous exception handler
+            set_unhandled_exception_handler(last)
+            last === nothing || last()
+            return
+        end
+
+        # setup a handler for unhandled exceptions (such as from root task or finish_task_hook)
+        # that tries to forward these errors to the REPL frontend
+        if isa(excpt, InterruptException) && backend.task.state == :runnable && isempty(Base.Workqueue)
```

Comment:
conservatism in minimizing the change in the conditional from the old version. i think the idea was to allow the backend task to try to unstick itself unless that seemed unlikely to happen on its own